src/bert/utils.py

```diff
@@ -0,0 +1,102 @@
+import logging as log
+
+from typing import Dict
+
+import torch
+import torch.nn as nn
+
+# huggingface implementation of BERT
+import pytorch_pretrained_bert
+
+class BertEmbedderModule(nn.Module):
+    """ Wrapper for BERT module to fit into jiant APIs. """
+
+    def __init__(self, args, cache_dir=None):
+        super(BertEmbedderModule, self).__init__()
+
+        self.model = \
+            pytorch_pretrained_bert.BertModel.from_pretrained(
+                args.bert_model_name,
+                cache_dir=cache_dir)
+        self.embeddings_mode = args.bert_embeddings_mode
+        # BERT supports up to 512 tokens; see section 3.2 of https://arxiv.org/pdf/1810.04805.pdf
+        assert args.max_seq_len <= 512
+        self.seq_len = args.max_seq_len
+
+        # Set trainability of this module.
+        for param in self.model.parameters():
+            param.requires_grad = bool(args.bert_fine_tune)
+
+    def forward(self, sent: Dict[str, torch.LongTensor],
+                unused_task_name: str="") -> torch.FloatTensor:
+        """ Run BERT to get hidden states.
+
+        Args:
+            sent: batch dictionary
+
+        Returns:
+            h: [batch_size, seq_len, d_emb]
+        """
+        assert "bert_wpm_pretokenized" in sent
+        # <int32> [batch_size, var_seq_len]
+        var_ids = sent["bert_wpm_pretokenized"]
+
+        # Model has fixed, learned positional component, so we must pass a
+        # block of exactly seq_len length.
+        # ids is <int32> [batch_size, seq_len]
+        ids = torch.zeros(var_ids.size()[0], self.seq_len, dtype=var_ids.dtype,
+                          device=var_ids.device)
+        fill_len = min(var_ids.size()[1], self.seq_len)
+        ids[:,:fill_len] = var_ids[:,:fill_len]
+
+        mask = (ids == 0)
+        # "Correct" ids to account for different indexing between BERT and
+        # AllenNLP.
+        # The AllenNLP indexer adds a '@@UNKNOWN@@' token to the
+        # beginning of the vocabulary, *and* treats that as index 1 (index 0 is
+        # reserved for padding).
+        FILL_ID = 0  # [PAD] for BERT models.
+        ids[ids == 0] = FILL_ID + 2
+        ids -= 2
```

Comment:
this will map elements of `ids` that are 1 to -1, potentially causing indexing errors; is there a guard against this or a reason it won't happen?

Reply:
Good catch. We pass the tokens through the BERT WordpieceTokenizer, which checks for OOV and replaces with it's own unk token https://github.com/huggingface/pytorch-pretrained-BERT/blob/master/pytorch_pretrained_bert/tokenization.py#L329 (which has some id > 1 in the main vocab), so this should never happen.

Added an assert to complain in case something goes wrong.

Reply:
sounds good, nit: should add a more descriptive error message here